.ci/ci-fast-return.sh

```diff
@@ -0,0 +1,426 @@
+	else
+		local_info "No repo set, skipping repo specific patterns"
+	fi
+
+	local canskip_exprs=""
+	for x in $common_skip $repo_skip; do
+		# Build up the string of "-e exp1 -e exp2" arguments to later pass
+		# to the egrep based search.
+		canskip_exprs+=$(echo "-e $x ")
+	done
+
+	local mustcheck_exprs=""
+	for x in $common_check $repo_check; do
+		# Build up the string of "-e exp1 -e exp2" arguments to later pass
+		# to the egrep based search.
+		mustcheck_exprs+=$(echo "-e $x ")
+	done
+
+	local_info "Skip patterns: [$canskip_exprs]"
+	local_info "Check patterns: [$mustcheck_exprs]"
+
+	# do we have any patterns to check?
+	if [ -n "$mustcheck_exprs" ]; then
+		local need_checking=$(check_matches ${mustcheck_exprs[@]})
+	else
+		local_info "No force CI check patterns"
+		local need_checking=""
+	fi
+
+	# If we have any files that *must* be checked, then immediately return
+	if [ -n "$need_checking" ]; then
+		# stderr, so it does not get in our return value...
+		cat >&2 <<-EOT
+		INFO: Cannot fastpath skip CI.
+		INFO: Some files present must be CI checked.
+		INFO: Files to check are:
+
+		$need_checking
+
+EOT
+		echo "1"
+		return 0
+	else
+		local_info "No force check files found"
+	fi
+
+	# Now we have checked there are no 'must check' files, we can check to see
+	# if all files fall into the 'can skip' patterns.
+	# first, do we have any skip patterns to search for?
+	if [ -n "$canskip_exprs" ]; then
+		local non_skippable=$(check_not_matches ${canskip_exprs[@]} <<< "$filenames")
+	else
+		local_info "No skip CI check patterns"
+		# No patterns, set non-skippable list to all files then...
+		local non_skippable="$filenames"
+	fi
+
+	if [ -z "$non_skippable" ]; then
+		# stderr, so it does not get in our return value...
+		cat >&2 <<-EOT
+		INFO: No files to check in CI.
+		INFO: Fastpath short circuit returning from CI.
+		INFO: Files skipped are:
+
+		$filenames
+
+EOT
+		echo "0"
+		return 0
+	else
+		cat >&2 <<-EOT
+		INFO: Not all files skippable
+
+		$non_skippable
+
+EOT
+		echo "1"
+		return 0
+	fi
+}
+
+# Check if we have the 'magic label' that forces a CI run set on the PR
+# Returns on stdout as string:
+#  0 - No label found, could skip the CI
+#  1 - Label found - should run the CI
+check_force_label() {
+	if [ -z "$ghprbGhRepository" ]; then
+		local_info "No ghprbGhRepository set, skip label check"
+		echo "0"
+		return 0
+	fi
+
+	if [ -z "$ghprbPullId" ]; then
+		local_info "No ghprbPullId set, skip label check"
+		echo "0"
+		return 0
+	fi
+
+	local_info "Checking labels for PR ${ghprbGhRepository}/${ghprbPullId}"
+
+	# Pull the label list for the PR
+	# Ideally we'd use a github auth token here so we don't get rate limited, but to do that we would
+	# have to expose the token into the CI scripts, which is then potentially a security hole.
+	local json=$(curl -sL https://api.github.com/repos/${ghprbGhRepository}/issues/${ghprbPullId}/labels)
+
+	# Pull the label list out
+	local labels=$(jq .[].name <<< $json)
```

Comment:
Nit: Should all the variables passed to `<<<` be quoted to minimise surprises?

Reply:
strangely... github passes all the labels back already enclosed in `"`s, which we then actually strip off a few lines further down in the loop before we do the string compare.... I think my head will explode if I try to contemplate what thus happens if we put a further set of `"`s around them here as well :-)

Reply:
We don't want :speaking_head: :boom:, so

```sh
cat <<< "lgtm ;)"
```